accord-core/src/main/java/accord/messages/ReadData.java

```diff
@@ -348,45 +380,42 @@ protected void onAllSuccess(@Nullable Ranges unavailable, @Nullable Data data, @
 
             case PENDING:
                 state = State.RETURNED;
-                node.reply(replyTo, replyContext, fail == null ? constructReadOk(unavailable, data) : null, fail);
+                node.reply(replyTo, replyContext, fail == null ? constructReadOk(unavailable, notReady, data) : null, fail);
                 break;
 
             default:
                 throw new AssertionError("Unknown state: " + state);
         }
     }
 
-    void beginCancel(@Nullable SafeCommandStore safeStore)
+    void cancel()
     {
-        if (safeStore != null)
-        {
-            int id = safeStore.commandStore().id();
-            cancelWaiting(safeStore);
-            waitingOn.clear(id);
-        }
+        if (timeout != null)
+            timeout.cancel();
 
-        // TODO (expected): efficient unsubscribe mechanism
-        node.commandStores().mapReduceConsume(this, waitingOn.stream(), forEach(safeStore0 -> cancelWaiting(safeStore0), node.agent()));
         state = State.OBSOLETE;
+        registrations.forEach((i, r) -> r.cancel());
         waitingOn.clear();
-        reading = null;
+        reading.clear();
         data = null;
         unavailable = null;
+        notReady = null;
     }
 
-    protected void beginWaiting(SafeCommandStore safeStore, boolean isExecuting)
+    synchronized public void timeout()
```

Comment:
Unused?

Reply:
it's invoked by `DefaultRequestTimeouts`?